ostd/src/mm/io.rs

```diff
@@ -536,9 +493,39 @@ impl<'a, Space> VmReader<'a, Space> {
         unsafe { self.cursor = self.cursor.add(nbytes) };
         self
     }
+
+    fn read_fallible_general<WF>(
```

Comment:
My intention is to provide a private function in `VmReader<F>` (maybe name `_read_fallible()` is better), and public APIs in block `impl VmReader<Fallible> {}` and `impl VmWriter<Infallible> {}` all invoke this private function. (No `read_fallible(writer: VmWriter<Infallible>)` in `VmReader<Infallible>` at all)

Since now the difference in this new version is we can pass both two types of writers to `read_fallible<F>(writer: VmWriter<F>)` in `VmReader<Fallible>`, note that there is an additional generic in this API while in `VmReader<Infallible>` donot. So the original macro can't work and the above common private function way came up.

I found that writing a new macro can be hard since two function signatures are different now. Here is an alternative approach:

```rust
pub trait FallibleRead<F> {
    /// Reads data into the writer given the fallibility.
    ///
    /// On success, the number of bytes read is returned;
    /// On error, both the error and the number of bytes read so far are returned.
    fn read_fallible(
        &mut self,
        writer: &mut VmWriter<'_, F>,
    ) -> core::result::Result<usize, (Error, usize)>;
}

pub trait FallibleWrite<F> {
    /// Writes data from the reader given the fallibility.
    ///
    /// On success, the number of bytes written is returned;
    /// On error, both the error and the number of bytes written so far are returned.
    fn write_fallible(
        &mut self,
        reader: &mut VmReader<'_, F>,
    ) -> core::result::Result<usize, (Error, usize)>;
}

macro_rules! impl_read_fallible {
    ($reader_fallibility:ty, $writer_fallibility:ty) => {
        impl<'a> FallibleRead<$writer_fallibility> for VmReader<'a, $reader_fallibility> {
            fn read_fallible(
                &mut self,
                writer: &mut VmWriter<'_, $writer_fallibility>,
            ) -> core::result::Result<usize, (Error, usize)> {
                /* omitted */
            }
        }
    };
}

impl_read_fallible!(Fallible, Infallible);
impl_read_fallible!(Fallible, Fallible);
impl_read_fallible!(Infallible, Fallible);
impl_write_fallible!(Fallible, Infallible);
impl_write_fallible!(Fallible, Fallible);
impl_write_fallible!(Infallible, Fallible);
```

A little complicated and we need to export these two new traits anywhere we use the fallible read/write, I kind of prefer the former "common private function" approach.